django-workload/django_workload/bundle_tray.py

```diff
@@ -0,0 +1,119 @@
+from .models import (
+    BundleEntryModel,
+    FeedEntryModel,
+    UserModel,
+)
+
+
+INC_FACTOR = 1
+
+
+class BundleTray(object):
+    def __init__(self, request):
+        self.request = request
+
+    def get_bundle(self):
+        bundles = list(
+            BundleEntryModel.objects
+            .filter(userid__in=self.request.user.following).limit(10))
+        # only one bundle per user
+        userids = {}
+        feedentryids = []
+        for bundle in bundles:
+            if bundle.userid in userids:
+                continue
+            userids[bundle.userid] = bundle.id
+            feedentryids += bundle.entry_ids
+        first_bundleids = set(userids.values())
+        # Fetch user information
+        userinfo = {}
+        for user in UserModel.objects.filter(id__in=list(userids)):
+            userinfo[user.id] = user.json_data
+        # fetch entry information
+        feedentryinfo = {}
+        for feedentry in FeedEntryModel.objects.filter(id__in=list(feedentryids)):
+            feedentryinfo[feedentry.id] = {
+                'pk': str(feedentry.id),
+                'comment_count': feedentry.comment_count,
+                'published': feedentry.published.timestamp(),
+            }
+
+        result = {'bundle': [
+            {
+                'pk': str(b.id),
+                'comment_count': b.comment_count,
+                'published': b.published.timestamp(),
+                'user': userinfo[b.userid],
+                'items': [
+                    feedentryinfo[f]
+                    for f in bundle.entry_ids if f in feedentryinfo]
+            }
+            for b in bundles if b.id in first_bundleids
+        ]}
+        return result
+
+    def get_inc_factor(self):
+        global INC_FACTOR
+        return int((INC_FACTOR + 1) / 2)
+
+    def post_process(self, res):
+        bundle_list = res['bundle']
+        conf = BundleConfig()
+
+        # duplicate the data
+        for i in range(conf.get_mult_factor()):
+            conf.list_extend(bundle_list)
+
+        sorted_list = sorted(conf.get_list(),
+                             key=lambda x: x['published'],
+                             reverse=True)
+        conf.final_items = []
+
+        for item in sorted_list:
+            conf.comm_total = conf.comm_total + item['comment_count']
+            for sub in item['items']:
+                conf.comm_total = conf.comm_total + sub['comment_count']
+            # un-duplicate the data
+            exists = False
+            for final_item in conf.final_items:
+                if final_item['published'] == item['published']:
+                    exists = True
+                    break
+            if not exists:
+                conf.final_items.append(item)
```

Comment:
This could be written without the need for `exists` as:
```py3
            for final_item in conf.final_items:
                if final_item['published'] == item['published']:
                    break
            else:
                conf.final_items.append(item)
```

Reply:
It doesn't look pretty, but it uses LOAD_FAST and STORE_FAST  more.